### PR TITLE
sql: support {GRANT|REVOKE} PRIVILEGES ... ON {SEQUENCE | ALL SEQUENCES IN SCHEMA}

### LINKED ISSUE #69584: sql: support GRANT ... ON ALL SEQUENCES IN SCHEMA ...
we currently support GRANT ON ALL TABLES IN SCHEMA which also grants on sequences but we should separate these out into two more granular commands.

Epic CRDB-14491
Jira issue: CRDB-9675

### LINKED ISSUE #74780: sql: support GRANT PRIVILEGES ... ON SEQUENCE
Sequences in PG have privileges.

[PG 13](https://www.postgresql.org/docs/13/sql-grant.html):

```

GRANT { { USAGE | SELECT | UPDATE }
    [, ...] | ALL [ PRIVILEGES ] }
    ON { SEQUENCE sequence_name [, ...]
         | ALL SEQUENCES IN SCHEMA schema_name [, ...] }
    TO role_specification [, ...] [ WITH GRANT OPTION ]

```

And also see this table:

Privilege | Abbreviation | Applicable Object Types
-- | -- | --
SELECT | r (“read”) | LARGE OBJECT, SEQUENCE, TABLE (and table-like objects), table column
INSERT | a (“append”) | TABLE, table column
UPDATE | w (“write”) | LARGE OBJECT, SEQUENCE, TABLE, table column
DELETE | d | TABLE
TRUNCATE | D | TABLE
REFERENCES | x | TABLE, table column
TRIGGER | t | TABLE
CREATE | C | DATABASE, SCHEMA, TABLESPACE
CONNECT | c | DATABASE
TEMPORARY | T | DATABASE
EXECUTE | X | FUNCTION, PROCEDURE
USAGE | U | DOMAIN, FOREIGN DATA WRAPPER, FOREIGN SERVER, LANGUAGE, SCHEMA, SEQUENCE, TYPE

This issue tracks whether we need to implement or put a placeholder for it or not, as it appears on PG dumps.

Jira issue: CRDB-12256

Epic CRDB-14491

### PR DESCRIPTION
**Pull request by ZhouXing19**
Add support for `{GRANT|REVOKE} PRIVILEGES ... ON {SEQUENCE | ALL SEQUENCES IN SCHEMA}` syntax.

Fixes #74780
Fixes #69584

Release Note(sql): Add support for `{GRANT|REVOKE} PRIVILEGES ... 
ON {SEQUENCE | ALL SEQUENCES IN SCHEMA}` syntax. 
The privileges allowed for sequences will be those in 
[Postgres14](https://www.postgresql.org/docs/14/sql-grant.html)
(`ALL, USAGE, SELECT, UPDATE`) in addition to the existing CRBD 
table privileges
(`ALL, CREATE, DROP, GRANT, SELECT, INSERT, DELETE, UPDATE,
 ZONECONFIG`).
This is because before 22.2 we allow granting table privileges to sequences
via `ALTER DEFAULT PRIVILEGES ... ON SEQUENCE`.
Note that now sequences are still treated as a subset of tables, 
which means `{GRANT|REVOKE} ... ON TABLE sequence_name` works,
 and `GRANT ... ON ALL TABLES IN SCHEMA`
also grant privileges on all sequences in the schema to the target user.

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
this looks quite nice! i just had a question

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-:-N-eCVpL-9jTcHwF0sKN:bwuusvc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @shermanCRL and @ZhouXing19)

___
*[pkg/sql/privilege/privilege.go, line 86 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N-eBcfk5WBzK32ZxDsL:-N-eBcfk5WBzK32ZxDsM:bwoakm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdfdcb6e8369bf8e1f7d33d2170d188f9ce9e2c8/pkg/sql/privilege/privilege.go#L86)):*
> ```Go
> 	SchemaPrivileges   = List{ALL, GRANT, CREATE, USAGE}
> 	TypePrivileges     = List{ALL, GRANT, USAGE}
> 	SequencePrivileges = List{ALL, USAGE, SELECT, UPDATE}
> ```

should we include all the TablePrivileges in the list for sequences? (since we are treating sequences as a special type of table)

otherwise, i'm a bit worried about the `catpb.IsValidPrivilegesForObjectType` and `privilege.ValidatePrivileges` functions
___
*[pkg/sql/logictest/testdata/logic_test/grant_sequence, line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N-eBLlH0blj1FRN0Czr:-N-eBLlI2LIv6pT8TlLF:bt2qtat) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdfdcb6e8369bf8e1f7d33d2170d188f9ce9e2c8/pkg/sql/logictest/testdata/logic_test/grant_sequence#L59)):*
> ```
> 
> statement ok
> GRANT ALL ON ALL SEQUENCES IN SCHEMA test.public TO readwrite;
> ```

could you add another test to see that `GRANT USAGE ON SEQUENCE` works? I'm a bit worried that since we still treat sequences the same as tables, then something somewhere will end up looking at `TablePrivileges`, which doesn't include `USAGE`

if USAGE doesn't work then i think it means we need to update the descriptor validation logic to be aware of sequence privileges.

**Review by ZhouXing19 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79862)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @shermanCRL)

___
*[pkg/sql/privilege/privilege.go, line 86 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N-eBcfk5WBzK32ZxDsL:-N-y8RXw9RTr9xdFh_8i:b-7xvtzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdfdcb6e8369bf8e1f7d33d2170d188f9ce9e2c8/pkg/sql/privilege/privilege.go#L86)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should we include all the TablePrivileges in the list for sequences? (since we are treating sequences as a special type of table)

otherwise, i'm a bit worried about the `catpb.IsValidPrivilegesForObjectType` and `privilege.ValidatePrivileges` functions
</blockquote></details>

I don't think we need to, because both of these 2 funcs call `privilge.GetValidPrivilegesForObject(objectType)` to return the allowed priv list here, and we refactored that function too, so that it always return this `SequencePrivileges` list when `objectType = priviledge.Sequence`.
___
*[pkg/sql/logictest/testdata/logic_test/grant_sequence, line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N-eBLlH0blj1FRN0Czr:-N-y4nz11mY9bqBFc_UF:b-oahk4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdfdcb6e8369bf8e1f7d33d2170d188f9ce9e2c8/pkg/sql/logictest/testdata/logic_test/grant_sequence#L59)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you add another test to see that `GRANT USAGE ON SEQUENCE` works? I'm a bit worried that since we still treat sequences the same as tables, then something somewhere will end up looking at `TablePrivileges`, which doesn't include `USAGE`

if USAGE doesn't work then i think it means we need to update the descriptor validation logic to be aware of sequence privileges.
</blockquote></details>

Yes good point, it didn't work because the `USAGE` privilege was not stored in the descriptor (`UserPrivileges.Privileges`).  I changed how we call `catprivilege.MaybeFixPrivileges` in `pkg/sql/catalog/tabledesc/table_desc_builder.go` and it's working now. See added tests above.

**Review by rafiss — CHANGES_REQUESTED** [review]
some small nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-:-N077bip196dWJtByKDA:baad9mi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @shermanCRL, and @ZhouXing19)

___
*[pkg/sql/sequence.go, line 102 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N076xZEDQ2Jkrekpesw:-N076xZEDQ2Jkrekpesx:bt37wwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f20ae1bcbbd61209fd27a3af534dafd5392d4674/pkg/sql/sequence.go#L102)):*
> ```Go
> 	ctx context.Context, p *planner, descriptor catalog.TableDescriptor,
> ) (int64, error) {
> 	var err error
> ```

nit: don't think we need the error declared here
___
*[pkg/sql/sequence.go, line 105 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N076rFt6nQyZDXi3CgQ:-N076rFt6nQyZDXi3CgR:bd4x8lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f20ae1bcbbd61209fd27a3af534dafd5392d4674/pkg/sql/sequence.go#L105)):*
> ```Go
> 
> 	requiredPrivileges := []privilege.Kind{privilege.USAGE, privilege.UPDATE}
> 	HasRequiredPriviledge := false
> ```

nit: should be lowercase `h`
___
*[pkg/sql/sequence.go, line 115 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N075tro9jnwgiy0SvUW:-N075trq39zfKNXvM9o8:bh3kdep) ([raw file](https://github.com/cockroachdb/cockroach/blob/f20ae1bcbbd61209fd27a3af534dafd5392d4674/pkg/sql/sequence.go#L115)):*
> ```Go
> 	}
> 	if !HasRequiredPriviledge {
> 		return 0, errors.AssertionFailedf(
> ```

nit: assertion error should only be used for programming errors. they cause reports to be sent to us and automatically  create github issues when it happens

so the error should match closer to what the `CheckPrivilege` function was returning:
```
	return pgerror.Newf(pgcode.InsufficientPrivilege,
		"user %s does not have %s privilege on %s %s",
		user, privilege, descriptor.DescriptorType(), descriptor.GetName())
```
___
*[pkg/sql/catalog/tabledesc/table_desc_builder.go, line 246 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N072BpuAaGCp_gOnpLE:-N072Bpw3NREMvJTZkoH:bi7cjqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f20ae1bcbbd61209fd27a3af534dafd5392d4674/pkg/sql/catalog/tabledesc/table_desc_builder.go#L246)):*
> ```Go
> 	}
> 
> 	fixedPrivileges := catprivilege.MaybeFixPrivileges(
> ```

i think the call to MaybeFixPrivileges will cause us to clear out any privileges that are now invalid for sequences (e.g. DELETE). i'm pretty sure that's fine, but just wanted to confirm. do you think we should call `runPostDeserializationChangesOnAllDescriptors` during the upgrade migration to v22.2? perhaps the schema team has thoughts

also, right now the comment on `MaybeFixPrivileges` says it can be removed after 21.2, but it seems like we should change that to say 22.2 now.

**Comment by ZhouXing19** [comment]
From the CI, it seems that the current changes are breaking default privilege tests. Looking into it.

**Review by ZhouXing19 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79862)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @shermanCRL)

___
*[`pkg/sql/sequence.go` line 102 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N076xZEDQ2Jkrekpesw:-N0WbeUBEibbPaed1dmj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f20ae1bcbbd61209fd27a3af534dafd5392d4674/pkg/sql/sequence.go#L102)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: don't think we need the error declared here
</blockquote></details>

Done.
___
*[`pkg/sql/sequence.go` line 105 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N076rFt6nQyZDXi3CgQ:-N0Wbgho6tQ3fXIr1dwQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f20ae1bcbbd61209fd27a3af534dafd5392d4674/pkg/sql/sequence.go#L105)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: should be lowercase `h`
</blockquote></details>

Done.
___
*[`pkg/sql/sequence.go` line 115 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N075tro9jnwgiy0SvUW:-N0WeGy39AXYmd6PNwYE:b-dtkds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f20ae1bcbbd61209fd27a3af534dafd5392d4674/pkg/sql/sequence.go#L115)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: assertion error should only be used for programming errors. they cause reports to be sent to us and automatically  create github issues when it happens

so the error should match closer to what the `CheckPrivilege` function was returning:
```
	return pgerror.Newf(pgcode.InsufficientPrivilege,
		"user %s does not have %s privilege on %s %s",
		user, privilege, descriptor.DescriptorType(), descriptor.GetName())
```
</blockquote></details>

Done, thanks!
___
*[`pkg/sql/catalog/tabledesc/table_desc_builder.go` line 246 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N072BpuAaGCp_gOnpLE:-N0WnbmB-d7agUtA-ylh:b-y000ae) ([raw file](https://github.com/cockroachdb/cockroach/blob/f20ae1bcbbd61209fd27a3af534dafd5392d4674/pkg/sql/catalog/tabledesc/table_desc_builder.go#L246)):*
> clear out any privileges that are now invalid for sequences (e.g. DELETE)

Yes that's the expected behavior.

> do you think we should call `runPostDeserializationChangesOnAllDescriptors` during the upgrade migration to v22.2?

Ah in this case I agree that we may want to include all the TablePrivileges in the list for sequences, so that we can avoid making existing privileges unavailable during upgrade migration. 
But I'm not sure we need to call `runPostDeserializationChangesOnAllDescriptors` since we're just expanding the set of allowed privileges for sequences. 

> also, right now the comment on MaybeFixPrivileges says it can be removed after 21.2, but it seems like we should change that to say 22.2 now.

Made the changes.

**Comment by rafiss** [comment]
> Ah in this case I agree that we may want to include all the TablePrivileges in the list for sequences, so that we can avoid making existing privileges unavailable during upgrade migration.

Could you say more? I thought you were saying that the privileges that were being removed are not used for sequences anywhere?

**Comment by ZhouXing19** [comment]
> I thought you were saying that the privileges that were being removed are not used for sequences anywhere?

They (e.g. `CREATE`) can be used for sequences via `ALTER DEFAULT PRIVILEGES`, e.g. https://github.com/cockroachdb/cockroach/blob/c67668928b7d9bbfe3df5d293397585dca4ca9d6/pkg/sql/logictest/testdata/logic_test/alter_default_privileges_with_grant_option#L554-L560

**Comment by ZhouXing19** [comment]
@rafiss re-pinging for review. Thanks!

**Review by rafiss — CHANGES_REQUESTED** [review]
all the code lgtm! i just have one comment about the notice.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-:-N2E9YDfB0XqsBkVNu0R:b5fgvor)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @shermanCRL, and @ZhouXing19)

___
*[`pkg/sql/grant_revoke.go` line 256 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N2E7F0iB7lDDp_-42Mg:-N2E7F0jEAkl2ng-6eq3:b-4xldls) ([raw file](https://github.com/cockroachdb/cockroach/blob/c73fae79c53edf51231afb3483c6a9d05b457586/pkg/sql/grant_revoke.go#L256)):*
> ```Go
> 					errors.WithHint(
> 						pgnotice.Newf("%s", sequencePrivilegesNotInPg.SortedNames()),
> 						"the above privileges may be incompatible with PostgreSQL",
> ```

i think we can use a bit of a softer message, so user's don't worry too much. also, i think we should only show a notice for the privileges that are essentially no-ops in CRDB. for example, even though DROP is not allowed in PG for sequences, it still is useful in CRDB. so i think the only privileges we need to show the notice for are: CREATE, INSERT, DELETE, and ZONECONFIG (but perhaps you could confirm that these all have no effect for sequences)

also, nit: no need to use `WithHint` on this one.

so, to summarize the above, i think we can use something like:

```
params.p.BufferClientNotice(
  ctx,
  pgnotice.Newf("some privileges have no effect on sequences: %s", sequencePrivilegesNotInPg.SortedNames(),
)
```

`List{ALL, USAGE, SELECT, UPDATE, CREATE, DROP, GRANT, INSERT, DELETE, ZONECONFIG}`

**Review by ZhouXing19 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79862)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @rafiss)

___
*[`pkg/sql/grant_revoke.go` line 256 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N2E7F0iB7lDDp_-42Mg:-N2Hp6If5bCnzxsOde8R:bvu9k5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c73fae79c53edf51231afb3483c6a9d05b457586/pkg/sql/grant_revoke.go#L256)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we can use a bit of a softer message, so user's don't worry too much. also, i think we should only show a notice for the privileges that are essentially no-ops in CRDB. for example, even though DROP is not allowed in PG for sequences, it still is useful in CRDB. so i think the only privileges we need to show the notice for are: CREATE, INSERT, DELETE, and ZONECONFIG (but perhaps you could confirm that these all have no effect for sequences)

also, nit: no need to use `WithHint` on this one.

so, to summarize the above, i think we can use something like:

```
params.p.BufferClientNotice(
  ctx,
  pgnotice.Newf("some privileges have no effect on sequences: %s", sequencePrivilegesNotInPg.SortedNames(),
)
```

`List{ALL, USAGE, SELECT, UPDATE, CREATE, DROP, GRANT, INSERT, DELETE, ZONECONFIG}`
</blockquote></details>

Thanks! This makes lots of sense. 
I renamed the list to `sequencePrivilegesNoOp`, and note that this excludes `DROP` and `GRANT`, which are not in PG's privileges for sequences but correspond to actual operations in CRDB. Test cases for `DROP` and `GRANT` on sequences are also added in `pkg/sql/logictest/testdata/logic_test/grant_sequence`.

**Review by rafiss — APPROVED** [review]
this looks great! feel free to merge after fixing the comment

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-:-N2Hr37zDXOjAHXt3XWL:b-u68str)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @rafiss, and @ZhouXing19)

___
*[`pkg/sql/privilege/privilege.go` line 89 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N2Hr0Yu5ByKoOKAKPG5:-N2Hr0Yu5ByKoOKAKPG6:b-vkkhzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b8c1fc68e0e9d8dd892f6883577e63f608f556/pkg/sql/privilege/privilege.go#L89)):*
> ```Go
> 	// before v22.2 we treated Sequences the same as Tables. This is to avoid making
> 	// certain privileges unavailable after upgrade migration.
> 	// Note that "UPDATE, CREATE, GRANT, "
> ```

nit: unfinished comment

**Comment by ZhouXing19** [comment]
*[`pkg/sql/grant_revoke.go` line 256 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/79862#-N2E7F0iB7lDDp_-42Mg:-N2Hs8717KaBA_21LDil:bn22wtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c73fae79c53edf51231afb3483c6a9d05b457586/pkg/sql/grant_revoke.go#L256)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Thanks! This makes lots of sense. 
I renamed the list to `sequencePrivilegesNoOp`, and note that this excludes `DROP` and `GRANT`, which are not in PG's privileges for sequences but correspond to actual operations in CRDB. Test cases for `DROP` and `GRANT` on sequences are also added in `pkg/sql/logictest/testdata/logic_test/grant_sequence`.
</blockquote></details>

And yes I've confirmed that`CREATE, INSERT, DELETE, ZONECONFIG` on sequences are no-ops.
